File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        occ = {}
        for num in nums:
            if num not in occ:
                occ[num] = 0
            occ[num] += 1
        n = len(nums)
        res = set()

        nums.sort()

        for i in range(n):
            for j in range(i):
                occ[nums[i]] -= 1
                occ[nums[j]] -= 1

                if -(nums[i] + nums[j]) in occ and occ[-(nums[i] + nums[j])] > 0:
                    res.add(tuple(sorted([nums[i], nums[j], -(nums[i] + nums[j])])))
                
                occ[nums[i]] += 1
                occ[nums[j]] += 1
        return list(res)

File: Data Structures & Algorithms/find-duplicate-integer/submission-2.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int n = (int)nums.size();
        for (int i = 0; i < n; i++) {
            if (nums[abs(nums[i]) - 1] < 0) {
                return abs(nums[i]);
            } else {
                nums[abs(nums[i]) - 1] *= -1;
            }
        }
        return 0;
    }
};
